### PR TITLE
support char mode emoji and multibyte emoji like 👨‍👩‍👧‍👦

### PR DESCRIPTION
**Pull request by ankye**
https://github.com/cocos/cocos-engine/issues/15192

Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/2d/utils/text-utils.ts [hunk 1]
```diff
@@ -214,6 +214,192 @@ export function safeMeasureText (ctx: CanvasRenderingContext2D, string: string,
...
+        if (charCode === 0x200d) {
+            len++;
+            continue;
+        }
+        if (charCode >= 0xd800 && charCode <= 0xdbff) {
+            len++;
+            charCode = str.charCodeAt(i + 1);
+            if (charCode >= 0xdc00 && charCode <= 0xdfff) {
+                if (i + 2 < length && str.charCodeAt(i + 2) === 0x200d) {
+                    len++;
```
Comment (qiuguohua) [inline cocos/2d/utils/text-utils.ts:234]: Maybe 'len++' can be placed above the if?

File: cocos/2d/utils/text-utils.ts [hunk 2]
```diff
@@ -214,6 +214,192 @@ export function safeMeasureText (ctx: CanvasRenderingContext2D, string: string,
...
+            return str.charAt(i);
+        }
+        start = i + 1;
+        count++;
+        len = 0;
+    }
+    return '';
+}
+
+export function getSymbolCodeAt (str: string, index: number): string  {
```
Comment (qiuguohua) [inline cocos/2d/utils/text-utils.ts:293]: ‘getSymbolAt’ and ‘getSymbolCodeAt’ have many similar codes. Can they share some codes?

File: cocos/2d/utils/text-utils.ts [hunk 3]
```diff
@@ -214,6 +214,192 @@ export function safeMeasureText (ctx: CanvasRenderingContext2D, string: string,
...
+                endCheckIndex++;
+                newEndIndex++;
+            }
+        }
+        endChar = targetString[endCheckIndex];
+        if (endChar >= '\uD800' && endChar <= '\uDBFF') {
+            // highSurrogateRex
+            endCheckIndex++;
+            newEndIndex++;
+            endChar = targetString[endCheckIndex];
```
Comment (qiuguohua) [inline cocos/2d/utils/text-utils.ts:387]: Can this be removed? Because the next loop will read

File: cocos/2d/utils/text-utils.ts [hunk 4]
```diff
@@ -214,6 +214,192 @@ export function safeMeasureText (ctx: CanvasRenderingContext2D, string: string,
...
+        } else if (endChar >= '\uDC00' && endChar <= '\uDFFF') {
+            // lowSurrogateRex
+            endCheckIndex++;
+            if (endCheckIndex < targetString.length && targetString[endCheckIndex] === '\u200d') {
+                newEndIndex++;
+                endChar = targetString[endCheckIndex];
```
Comment (qiuguohua) [inline cocos/2d/utils/text-utils.ts:393]: Same as above

**Comment by qiuguohua** [comment]
Is it possible to add some unit tests?

**Comment by ankye** [comment]
> Is it possible to add some unit tests?
https://github.com/airkits/toolsTP/blob/main/src/test-emoji.ts
you can clone this project and test
npm install
ts-node ./src/test-emoji.ts

**Comment by ankye** [comment]
> Is it possible to add some unit tests?

let txt = 'w大熊🐻👨‍👩‍👧‍👦🐼'
console.log(txt, txt.length)
// console.log(getSymbolStartIndex(txt, 0), 0)
// console.log(getSymbolStartIndex(txt, 1), 1)
// console.log(getSymbolStartIndex(txt, 2), 2)
// console.log(getSymbolStartIndex(txt, 3), 3)
// console.log(getSymbolStartIndex(txt, 4), 3)
// console.log(getSymbolStartIndex(txt, 5), 5)
// console.log(getSymbolStartIndex(txt, 6), 5)
// console.log(getSymbolStartIndex(txt, 7), 5)
// console.log(getSymbolStartIndex(txt, 8), 5)
// console.log(getSymbolStartIndex(txt, 9), 5)
// console.log(getSymbolStartIndex(txt, 10), 5)
// console.log(getSymbolStartIndex(txt, 11), 5)
// console.log(getSymbolStartIndex(txt, 12), 5)
// console.log(getSymbolStartIndex(txt, 13), 5)
// console.log(getSymbolStartIndex(txt, 14), 5)
// console.log(getSymbolStartIndex(txt, 15), 5)
// console.log(getSymbolStartIndex(txt, 16), 16)
// console.log(getSymbolStartIndex(txt, 17), 16)

// console.log(getSymbolEndIndex(txt, 0), 0)
// console.log(getSymbolEndIndex(txt, 1), 1)
// console.log(getSymbolEndIndex(txt, 2), 2)
// console.log(getSymbolEndIndex(txt, 3), 4)
// console.log(getSymbolEndIndex(txt, 4), 4)
// console.log(getSymbolEndIndex(txt, 5), 15)
// console.log(getSymbolEndIndex(txt, 6), 15)
// console.log(getSymbolEndIndex(txt, 7), 15)
// console.log(getSymbolEndIndex(txt, 8), 15)
// console.log(getSymbolEndIndex(txt, 9), 15)
// console.log(getSymbolEndIndex(txt, 10), 15)
// console.log(getSymbolEndIndex(txt, 11), 15)
// console.log(getSymbolEndIndex(txt, 12), 15)
// console.log(getSymbolEndIndex(txt, 13), 15)
// console.log(getSymbolEndIndex(txt, 14), 15)
// console.log(getSymbolEndIndex(txt, 15), 15)
// console.log(getSymbolEndIndex(txt, 16), 17)
// console.log(getSymbolEndIndex(txt, 17), 17)

//console.log(_safeSubstring(txt, 0, 6), '&&', _safeSubstring(txt, 7, 15), '&&', _safeSubstring(txt, 16, 17))

// console.log(_safeSubstring(txt, 0, 3), '&&', _safeSubstring(txt, 4, 5), '&&', _safeSubstring(txt, 6, 17))
// console.log(_safeSubstring(txt, 0, 2), '&&', _safeSubstring(txt, 3, 8), '&&', _safeSubstring(txt, 9, 17))

**Comment by zhefengzhang** [comment]
This PR solves the issue https://github.com/cocos/cocos-engine/issues/16710 .

**Comment by qiuguohua** [comment]
@ankye Can you help me reply to the review suggestions? If you don't have time, can I help improve it?

**Comment by ankye** [comment]
> @ankye Can you help me reply to the review suggestions? If you don't have time, can I help improve it?

Yeah, I've been a little busy lately
you can modify and improve

File: cocos/2d/utils/text-utils.ts [hunk 5]
```diff
@@ -224,27 +410,25 @@ export function safeMeasureText (ctx: CanvasRenderingContext2D, string: string,
...
-                newEndIndex--;
-            }
-        } else if (startChar >= '\uD800' && startChar <= '\uDBFF') {
-            // highSurrogateRex
-            newEndIndex++;
+        endIndex = Math.max(0, endIndex - 1);
+        newEndIndex = getSymbolEndIndex(targetString, endIndex);
+        const newStartEndIndex = getSymbolStartIndex(targetString, endIndex);
+        if (newStartEndIndex < newStartIndex || (newStartEndIndex == newStartIndex && startIndex > newStartIndex)) {
+            newEndIndex = newStartIndex;
```
Comment (qiuguohua) [inline cocos/2d/utils/text-utils.ts:424]: This judgment is entered when '_safeSubstring(txt, 4, 3)' is called, but the empty string is returned. I don't understand the purpose of this? It should return an error?

Comment (ankye) [reply cocos/2d/utils/text-utils.ts:424]: startIndex > endIndex is not considered，you can help me to add this judgment

the multi-byte substring needs to consider the substrings concatenated same as the original string，The way I do this is I pick the first byte as whether in the substring，For example
_safeSubstring(txt, 0, 3) + _safeSubstring(txt, 4, 5) + _safeSubstring(txt, 6, 17)) 
=_safeSubstring(txt, 0, 17)
so if startindex is not the first byte，and endindex and startindex as the same char，should return empty string

Comment (qiuguohua) [reply cocos/2d/utils/text-utils.ts:424]: Okay, got it.
I've finished optimizing it and can help check it when you have time
https://github.com/cocos/cocos-engine/pull/16738

**Pull request closed without merging**